My take on this week's homework

CapnKernel opened the pull request:
Hi!

This is where I'm at with the homework.

I'm a bit concerned that the README.md file implies multiple models for normalisation.  The only one that makes sense to me is to pull out client.  Can you please suggest additional targets for normalisation?

Thanks!

Commit:
My take on this week's homework

I'm a bit concerned that the README.md file implies multiple models for
normalisation.

funkybob commented on timetracker/entries/migrations/0003_move_project_client_to_model.py:
```diff
@@ -0,0 +1,84 @@
+        except Client.DoesNotExist:
+            # No, so create it
+            c = Client(name=project.client_as_str)
+            c.save()
```
These two steps could have been one using Client.objects.create()

funkybob commented on timetracker/entries/migrations/0003_move_project_client_to_model.py:
```diff
@@ -0,0 +1,84 @@
+    # Then set the client property to be the new or existing Client.
+    for project in Project.objects.all():
+        # Is there an existing client of this name?
+        try:
```
This try/except could have been done using Client.objects.get_or_create()

Commit:
Switch to get_or_create(), it's clearer.

Commit:
Set text client to allow null

Even though it will be deleted shortly, it needs to allow null so that
the reverse migration works.  (It breaks when the column is added, and
there's no default value and null's not allowed).

sesh commented on README.md:
```diff
@@ -10,12 +10,12 @@ https://melbdjango.github.io/lesson-two/
-- [ ] [Fork this repository][gh-fork]
-- [ ] Clone the repo to your own machine
-- [ ] Normalise the model we created in class by creating new [Django models][dj-models] that better express the
+- [X] [Fork this repository][gh-fork]
```
Checking things off!

sesh commented on timetracker/entries/migrations/0003_move_project_client_to_model.py:
```diff
@@ -0,0 +1,83 @@
+        migrations.AlterField(
+            model_name='project',
+            name='client',
+            field=models.CharField(max_length=200, null=True),
```
Is there a specific reason that you're making this field nullable?

sesh commented on timetracker/entries/migrations/0004_add_entry_pluralisation.py:
```diff
@@ -0,0 +1,18 @@
+    operations = [
+        migrations.AlterModelOptions(
+            name='entry',
+            options={'verbose_name_plural': 'entries'},
```
Even more bonus points!

sesh commented on timetracker/entries/models.py:
```diff
@@ -2,24 +2,40 @@
 class Project(models.Model):
-    client = models.CharField(max_length=200)
     name = models.CharField(max_length=200)
+    client = models.ForeignKey(Client)
```
Little heads up: it's safer to put the model name in quotes in `ForeignKey` and `ManyToMany` field definitions.

sesh commented:
Nice work @CapnKernel.
